docs/design-documents/20240812-recover-from-pipeline-errors.md

```diff
@@ -0,0 +1,194 @@
+# Recover from Pipeline Errors
+
+## Introduction
+
+### Goals
+
+The primary goal of this design document is to introduce a robust error-handling
+mechanism within Conduit to ensure the reliability and resilience of data
+pipelines. The objectives include:
+
+- **Error Classification:** Distinguish between transient (recoverable) and
+  fatal (non-recoverable) errors.
+- **Automatic Recovery:** Implement a mechanism to automatically restart
+  pipelines in the event of transient errors with configurable backoff
+  strategies.
+- **Pipeline State Management:** Introduce new pipeline states to clearly
+  indicate the status of recovery processes.
+- **Configurable Backoff Strategy:** Allow users to configure backoff parameters
+  to suit different production environments.
+
+## Background
+
+Conduit's engine is currently very strict, stopping the pipeline and putting it
+into a degraded state whenever an error occurs. This approach is not optimal for
+production environments where transient errors, such as connectivity issues, may
+cause frequent disruptions. To enhance reliability, Conduit should include a
+mechanism to automatically restart failing pipelines, leveraging the fact that
+Conduit already tracks acknowledged record positions, making restarts a safe
+operation with no risk of data loss (assuming connectors are implemented
+correctly).
+
+## Error classification
+
+### Fatal errors
+
+Fatal errors are defined as non-recoverable issues that require manual
+intervention. When a fatal error occurs, the pipeline will be stopped
+permanently, and no further attempts will be made to restart it. This approach
+ensures that critical issues are addressed promptly and prevents Conduit from
+entering an endless loop of restarts.
+
+### Transient errors
+
+Transient errors are recoverable issues that do not require manual intervention.
+Examples include temporary network failures or timeouts. For such errors,
+Conduit will automatically attempt to restart the pipeline using a backoff delay
+strategy.
+
+- **Default Behavior:** By default, all errors are considered transient and
+  recoverable. A connector developer must explicitly mark an error as fatal if
+  it should cause the pipeline to stop permanently.
+
+## Recovery mechanism
+
+### Backoff delay strategy
+
+To manage transient errors and prevent constant pipeline restarts, Conduit will
+implement a backoff delay strategy. This strategy introduces a waiting period
+before attempting to restart the pipeline, allowing time for issues to resolve
+themselves. The backoff parameters will be configurable to allow customization
+based on specific production needs.
+
+- **Configurable Parameters:**
+  - **Minimum Delay Before Restart:** Default: 1 second
+  - **Maximum Delay Before Restart:** Default: 1 minute
+  - **Backoff Factor:** Default: 2
+  - **Maximum Number of Retries:** Default: Infinite
+  - **Delay After Which Fail Count is Reset:** Default: 5 minutes (or on first
```

Comment:
I'm not sure I understand this part... Does this mean that after 5 minutes, we start counting the failures from 0?

Reply:
> pipeline is running for at least 5 minutes, i.e. that's when we consider it healthy again

Yes, the purpose of this is that a pipeline might be healthy even if no data is flowing, since the source just may have no records to produce. In that case, we would mark the pipeline as healthy after some period of time.

This is not ideal of course, but unless we add a specific "health" method on the connector to tell us if everything is fine, we have no way of knowing if a pipeline is doing fine or still trying to recover.

We could just as well skip this, stay on the cautious side and leave the pipeline in `recovering` until we get the first record, although in that case a healthy pipeline could be marked as `recovering` for a potentially long time.